File: core/planner/src/main/java/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMdHetDistribution.java
package ch.epfl.dias.calcite.adapter.pelago.metadata;

import org.apache.calcite.plan.volcano.RelSubset;
import org.apache.calcite.rel.BiRel;
import org.apache.calcite.rel.RelDistribution;
import org.apache.calcite.rel.RelDistributionTraitDef;
import org.apache.calcite.rel.RelDistributions;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.SingleRel;
import org.apache.calcite.rel.core.Project;
import org.apache.calcite.rel.metadata.BuiltInMetadata;
import org.apache.calcite.rel.metadata.ChainedRelMetadataProvider;
import org.apache.calcite.rel.metadata.MetadataDef;
import org.apache.calcite.rel.metadata.MetadataHandler;
import org.apache.calcite.rel.metadata.ReflectiveRelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMdDistribution;
import org.apache.calcite.rel.metadata.RelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMetadataQuery;
import org.apache.calcite.rex.RexNode;
import org.apache.calcite.util.BuiltInMethod;
import org.apache.calcite.util.mapping.Mappings;

import com.google.common.collect.ImmutableList;

import ch.epfl.dias.calcite.adapter.pelago.PelagoDeviceCross;
import ch.epfl.dias.calcite.adapter.pelago.PelagoSplit;
import ch.epfl.dias.calcite.adapter.pelago.PelagoTableScan;
import ch.epfl.dias.calcite.adapter.pelago.PelagoUnion;
import ch.epfl.dias.calcite.adapter.pelago.PelagoUnpack;
import ch.epfl.dias.calcite.adapter.pelago.RelHetDistribution;
import ch.epfl.dias.calcite.adapter.pelago.RelHetDistributionTraitDef;

import java.util.List;

public class PelagoRelMdHetDistribution implements MetadataHandler<HetDistribution> {
  private static final PelagoRelMdHetDistribution INSTANCE = new PelagoRelMdHetDistribution();

  public static final RelMetadataProvider SOURCE = ReflectiveRelMetadataProvider.reflectiveSource(
          HetDistribution.method, PelagoRelMdHetDistribution.INSTANCE
  );

  public MetadataDef<HetDistribution> getDef() {
    return HetDistribution.DEF;
  }

  public RelHetDistribution hetDistribution(RelNode rel, RelMetadataQuery mq) {
    return RelHetDistributionTraitDef.INSTANCE.getDefault();
  }

  public RelHetDistribution hetDistribution(RelSubset rel, RelMetadataQuery mq) {
    return rel.getTraitSet().getTrait(RelHetDistributionTraitDef.INSTANCE);
  }

  public RelHetDistribution hetDistribution(SingleRel rel, RelMetadataQuery mq) {
    return ((PelagoRelMetadataQuery) mq).hetDistribution(rel.getInput());
  }

  public RelHetDistribution hetDistribution(PelagoUnpack rel, RelMetadataQuery mq) {
    return ((PelagoRelMetadataQuery) mq).hetDistribution(rel.getInput());
  }

  public RelHetDistribution hetDistribution(BiRel rel, RelMetadataQuery mq) {
    return ((PelagoRelMetadataQuery) mq).hetDistribution(rel.getRight());
  }

  public RelHetDistribution hetDistribution(PelagoTableScan scan, RelMetadataQuery mq){
    return RelHetDistribution.SINGLETON;
  }

  public RelHetDistribution hetDistribution(PelagoSplit split, RelMetadataQuery mq) {
    return split.hetdistribution();
  }

  public RelHetDistribution hetDistribution(PelagoUnion union, RelMetadataQuery mq) {
    return RelHetDistribution.SINGLETON;
  }

}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMdHetDistribution.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.metadata

import ch.epfl.dias.calcite.adapter.pelago.rel.{
  PelagoSplit,
  PelagoTableScan,
  PelagoUnion,
  PelagoUnpack
}
import ch.epfl.dias.calcite.adapter.pelago.traits.{
  RelHetDistribution,
  RelHetDistributionTraitDef
}
import org.apache.calcite.plan.volcano.RelSubset
import org.apache.calcite.rel.{BiRel, RelNode, SingleRel}
import org.apache.calcite.rel.metadata.{
  MetadataDef,
  MetadataHandler,
  ReflectiveRelMetadataProvider,
  RelMetadataProvider,
  RelMetadataQuery
}

object PelagoRelMdHetDistribution {
  private val INSTANCE = new PelagoRelMdHetDistribution
  val SOURCE: RelMetadataProvider =
    ReflectiveRelMetadataProvider.reflectiveSource(
      HetDistribution.method,
      PelagoRelMdHetDistribution.INSTANCE
    )
}
class PelagoRelMdHetDistribution extends MetadataHandler[HetDistribution] {
  override def getDef: MetadataDef[HetDistribution] = HetDistribution.DEF

  def hetDistribution(rel: RelNode, mq: RelMetadataQuery): RelHetDistribution =
    RelHetDistributionTraitDef.INSTANCE.getDefault

  def hetDistribution(
      rel: RelSubset,
      mq: RelMetadataQuery
  ): RelHetDistribution =
    rel.getTraitSet.getTrait(RelHetDistributionTraitDef.INSTANCE)

  def hetDistribution(
      rel: SingleRel,
      mq: RelMetadataQuery
  ): RelHetDistribution =
    mq.asInstanceOf[PelagoRelMetadataQuery].hetDistribution(rel.getInput)

  def hetDistribution(
      rel: PelagoUnpack,
      mq: RelMetadataQuery
  ): RelHetDistribution =
    mq.asInstanceOf[PelagoRelMetadataQuery].hetDistribution(rel.getInput)

  def hetDistribution(rel: BiRel, mq: RelMetadataQuery): RelHetDistribution =
    mq.asInstanceOf[PelagoRelMetadataQuery].hetDistribution(rel.getRight)

  def hetDistribution(
      scan: PelagoTableScan,
      mq: RelMetadataQuery
  ): RelHetDistribution =
    RelHetDistribution.SINGLETON

  def hetDistribution(
      split: PelagoSplit,
      mq: RelMetadataQuery
  ): RelHetDistribution =
    split.hetdistribution

  def hetDistribution(
      union: PelagoUnion,
      mq: RelMetadataQuery
  ): RelHetDistribution =
    RelHetDistribution.SINGLETON
}
